boltons/ioutils.py

```diff
@@ -0,0 +1,439 @@
+        if not self._rolled:
+            tmp = TemporaryFile()
+            pos = self.buffer.tell()
+            tmp.write(self.buffer.getvalue())
+            tmp.seek(pos)
+            self.buffer.close()
+            self._buffer = tmp
+
+    @property
+    def _rolled(self):
+        return not isinstance(self.buffer, BytesIO)
+
+    @property
+    def buffer(self):
+        try:
+            return self._buffer
+        except AttributeError:
+            self._buffer = BytesIO()
+        return self._buffer
+
+    @property
+    def len(self):
+        """Determine the length of the file"""
+        pos = self.tell()
+        if self._rolled:
+            self.seek(0)
+            val = os.fstat(self.fileno()).st_size
+        else:
+            self.seek(0, os.SEEK_END)
+            val = self.tell()
+        self.seek(pos)
+        return val
+
+    def tell(self):
+        return self.buffer.tell()
+
+
+class SpooledStringIO(SpooledIOBase):
+    """
+    SpooledStringIO is a spooled file-like-object that only accepts unicode
+    values. On Python 2.x this means the 'unicode' type and on Python 3.x this
+    means the 'str' type. Values are accepted as unicode and then coerced into
+    utf-8 encoded bytes for storage. On retrieval, the values are returned as
+    unicode.
+
+    Example::
+
+        >>> from boltons import ioutils
+        >>> with ioutils.SpooledStringIO() as f:
+        ...     f.write(u"\u2014 Hey, an emdash!")
+        ...     _ = f.seek(0)
+        ...     isinstance(f.read(), ioutils.text_type)
+        True
+
+     """
+
+    def read(self, n=-1):
+        return self.buffer.read(n).decode('utf-8')
+
+    def write(self, s):
+        if not isinstance(s, text_type):
+            raise TypeError("{0} expected, got {1}".format(
+                text_type.__name__,
+                type(s).__name__
+            ))
+
+        self.buffer.write(s.encode('utf-8'))
+        if self.tell() >= self._max_size:
+            self.rollover()
+
+    def _traverse_codepoints(self, current_position, n):
+        """Traverse from current position to the right n codepoints"""
+        dest = current_position + n
+        while True:
+            if current_position == dest:
+                # By chance we've landed on the right position, break
+                break
+
+            # If the read would take us past the intended position then
+            # seek only enough to cover the offset
+            if current_position + READ_CHUNK_SIZE > dest:
+                self.read(dest - current_position)
+                break
+            else:
+                ret = self.read(READ_CHUNK_SIZE)
+
+            # Increment our current position
+            current_position += READ_CHUNK_SIZE
+
+            # If we kept reading but there was nothing here, break
+            # as we are at the end of the file
+            if not ret:
+                break
+
+        return dest
+
+    def seek(self, pos, mode=0):
```

Comment:
I wrote a small [Hypotheis](http://hypothesis.readthedocs.io/en/latest/) test case for `seek` with `SEEK_SET`:

```python
import pytest
import io

from boltons import ioutils
from hypothesis import strategies as st, given


def make_spooled_stringio():
    return ioutils.SpooledStringIO(32)


@given(text=st.text(),
       data=st.data())
@pytest.mark.parametrize('make_sio', [
    io.StringIO,
    make_spooled_stringio,
])
def test_seek(text, data, make_sio):
    where = data.draw(st.integers(min_value=0, max_value=len(text)))
    s = make_sio()
    s.write(text)
    s.seek(where)
    assert s.read() == text[where:]
```

The invariant under test is that seeking forward from the beginning any amount between 0 and the length of the string and then reading the remainder should return a string equal to the slice beginning there.

All Hypothesis test data passes with `io.StringIO`, but it fails with `SpooledStringIO` and the following inputs:

```python
sio = SpooledStringIO(32)
sio.write(u'\x8000')
sio.seek(2)
sio.read()  # prints u'', not u'0'
```

This appears to be unexpected behavior courtesy of `EncodedFile`

```
>>> c = EncodedFile(io.BytesIO(), data_encoding='utf-8')
>>> c.write(u'\x8000'.encode('utf-8'))
>>> c.seek(0)
0L
>>> len(c.read(2).decode('utf-8'))
3
````

That is, a request to read 2 code points actually returns 3!

It's certainly true that *your* code is on point!  But given the difficulty presented by `EncodingFile`, It occurs to me that it may be worthwhile to take a different approach.  What if you used `utf-32` instead of `utf-8`?  Because UTF-32 is a fixed-width encoding, you wouldn't need `EncodedFile`, but instead could just do modulo 4 arithmetic for seeks and tells.  It's worth noting that CPython itself uses one of several fixed-width representations of Unicode strings to avoid the hairy code you've successfully written and also for the sake of greater efficiency; with a fixed width encoding you can seek backwards from the end without having to go forward from the beginning!